File: web/src/map/geo.ts
export type LonLat = [number, number];

const R = 6371;

export function haversineKm(a: LonLat, b: LonLat): number {
  const toRad = (d: number) => (d * Math.PI) / 180;
  const dLat = toRad(b[1] - a[1]);
  const dLon = toRad(b[0] - a[0]);
  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(a[1])) * Math.cos(toRad(b[1])) * Math.sin(dLon / 2) ** 2;
  return 2 * R * Math.asin(Math.sqrt(h));
}

/** Делит ломаную на пройденную и оставшуюся части по доле длины. */
export function splitTrack(track: LonLat[], fraction: number): { done: LonLat[]; rest: LonLat[] } {
  if (track.length < 2) return { done: track, rest: [] };
  const total = track.slice(1).reduce((acc, p, i) => acc + haversineKm(track[i], p), 0);
  const target = Math.min(Math.max(fraction, 0), 1) * total;
  const done: LonLat[] = [track[0]];
  let walked = 0;
  for (let i = 0; i < track.length - 1; i++) {
    const seg = haversineKm(track[i], track[i + 1]);
    if (walked + seg >= target) {
      const t = seg > 0 ? (target - walked) / seg : 0;
      const split: LonLat = [
        track[i][0] + (track[i + 1][0] - track[i][0]) * t,
        track[i][1] + (track[i + 1][1] - track[i][1]) * t,
      ];
      done.push(split);
      return { done, rest: [split, ...track.slice(i + 1)] };
    }
    walked += seg;
    done.push(track[i + 1]);
  }
  return { done, rest: [] };
}

File: app/services/geo.py
"""Геометрия на сфере: расстояния, азимуты, интерполяция вдоль ломаной.

Координаты везде — (lat, lon) в градусах; в GeoJSON порядок обратный
([lon, lat]) — конвертируется на границе API, не здесь.
"""

import math
from collections.abc import Sequence

EARTH_RADIUS_KM = 6371.0

LatLon = tuple[float, float]


def haversine_km(a: LatLon, b: LatLon) -> float:
    """Расстояние по дуге большого круга, км."""
    lat1, lon1 = map(math.radians, a)
    lat2, lon2 = map(math.radians, b)
    d_lat = lat2 - lat1
    d_lon = lon2 - lon1
    h = math.sin(d_lat / 2) ** 2 + math.cos(lat1) * math.cos(lat2) * math.sin(d_lon / 2) ** 2
    return 2 * EARTH_RADIUS_KM * math.asin(math.sqrt(h))


def bearing_deg(a: LatLon, b: LatLon) -> float:
    """Начальный азимут из a в b, градусы по часовой от севера (0..360)."""
    lat1, lon1 = map(math.radians, a)
    lat2, lon2 = map(math.radians, b)
    d_lon = lon2 - lon1
    x = math.sin(d_lon) * math.cos(lat2)
    y = math.cos(lat1) * math.sin(lat2) - math.sin(lat1) * math.cos(lat2) * math.cos(d_lon)
    return (math.degrees(math.atan2(x, y)) + 360.0) % 360.0


def polyline_length_km(points: Sequence[LatLon]) -> float:
    return sum(haversine_km(points[i], points[i + 1]) for i in range(len(points) - 1))


def point_along(points: Sequence[LatLon], fraction: float) -> tuple[LatLon, float]:
    """Точка на ломаной на доле fraction (0..1) её длины и курс в этой точке.

    Интерполяция линейная по сегментам — для отрисовки и оценки положения
    этого достаточно (сегменты короткие относительно радиуса Земли).
    """
    if not points:
        raise ValueError("пустая ломаная")
    if len(points) == 1:
        return points[0], 0.0
    fraction = min(max(fraction, 0.0), 1.0)
    total = polyline_length_km(points)
    if total == 0:
        return points[0], 0.0
    target = fraction * total
    walked = 0.0
    for i in range(len(points) - 1):
        seg = haversine_km(points[i], points[i + 1])
        if walked + seg >= target or i == len(points) - 2:
            t = (target - walked) / seg if seg > 0 else 0.0
            t = min(max(t, 0.0), 1.0)
            lat = points[i][0] + (points[i + 1][0] - points[i][0]) * t
            lon = points[i][1] + (points[i + 1][1] - points[i][1]) * t
            return (lat, lon), bearing_deg(points[i], points[i + 1])
        walked += seg
    return points[-1], bearing_deg(points[-2], points[-1])


def point_in_polygon(point: LatLon, polygon: Sequence[LatLon]) -> bool:
    """Лежит ли точка внутри многоугольника (лучевой метод, плоское приближение).

    Для морских полигонов масштаба сотен километров искажение проекции
    (lat, lon) как плоскости не важно: границы и так грубые.
    """
    lat, lon = point
    inside = False
    n = len(polygon)
    for i in range(n):
        lat_a, lon_a = polygon[i]
        lat_b, lon_b = polygon[(i + 1) % n]
        if (lat_a > lat) == (lat_b > lat):
            continue  # ребро не пересекает горизонталь через точку
        cross_lon = lon_a + (lat - lat_a) * (lon_b - lon_a) / (lat_b - lat_a)
        if lon < cross_lon:
            inside = not inside
    return inside
